src/dm/DispatchManagerActions.cc

```diff
@@ -2256,3 +2469,751 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    vmm->trigger_backup_cancel(vid);
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int DispatchManager::restore(int vid, int img_id, int inc_id, int disk_id,
+            const RequestAttributes& ra, std::string& error_str)
+{
+    auto vm = vmpool->get(vid);
+
+    if ( vm == nullptr )
+    {
+        error_str = "Could not restore VM, it does not exist";
+        return -1;
+    }
+
+    if (vm->get_state() != VirtualMachine::POWEROFF)
+    {
+        error_str = "Could not restore VM, it must be in poweroff";
+        return -1;
+    }
+
+    vm->set_state(VirtualMachine::ACTIVE);
+    vm->set_state(VirtualMachine::PROLOG_RESTORE);
+
+    // Call driver action to copy disk from image backup to VM system disk
+    tm->trigger_prolog_restore(vm->get_oid(), img_id, inc_id, disk_id);
+
+    vm->set_vm_info();
+
+    close_cp_history(vmpool, vm.get(), VMActions::RESTORE_ACTION, ra);
+
+    vmpool->update(vm.get());
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+static int test_set_capacity(VirtualMachine * vm, float cpu, long mem, int vcpu,
+        string& error)
+{
+    HostPool * hpool = Nebula::instance().get_hpool();
+
+    int rc;
+
+    if ( vm->get_state() == VirtualMachine::POWEROFF ||
+         (vm->get_state() == VirtualMachine::ACTIVE &&
+          vm->get_lcm_state() == VirtualMachine::RUNNING))
+    {
+        HostShareCapacity sr;
+
+        auto host = hpool->get(vm->get_hid());
+
+        if ( host == nullptr )
+        {
+            error = "Could not update host";
+            return -1;
+        }
+
+        vm->get_capacity(sr);
+
+        host->del_capacity(sr);
+
+        rc = vm->resize(cpu, mem, vcpu, error);
+
+        if ( rc == -1 )
+        {
+            return -1;
+        }
+
+        vm->get_capacity(sr);
+
+        if (!host->test_capacity(sr, error))
+        {
+            return -1;
+        }
+
+        host->add_capacity(sr);
+
+        hpool->update(host.get());
+    }
+    else
+    {
+        rc = vm->resize(cpu, mem, vcpu, error);
+    }
+
+    return rc;
+}
+
+int DispatchManager::resize(int vid, float cpu, int vcpu, long memory,
+                            const RequestAttributes& ra, string& error_str)
+{
+    /* ---------------------------------------------------------------------- */
+    /*  Check & update VM & host capacity                                     */
+    /* ---------------------------------------------------------------------- */
+    auto vm = vmpool->get(vid);
+
+    if (vm == nullptr)
+    {
+        error_str = "Virtual machine does not exist";
+
+        return -1;
+    }
+
+    int rc;
+
+    switch (vm->get_state())
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The switch statement here is missing a default statement. The default statement will catch any case values that are not explicitly handled by a case statement.  This prevents bugs if a case value is accidentally forgotten or new enum values are added later